### PR TITLE
Add endpoints to talk to cleezy and frontend to add, delete, and list urls

### PR DESCRIPTION
**Pull request by pnavab**

### CODE REVIEW

File: api/main_endpoints/routes/URLShortener-Endpoints.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
+const express = require('express');
+const axios = require('axios');
+const router = express.Router();
+
+let URL_SHORTENER_BASE_URL = 'http://localhost:8000';
```
Comment (evanugarte) [inline api/main_endpoints/routes/URLShortener-Endpoints.js:5]: what about when this actually runs on the sce website?

File: api/main_endpoints/routes/URLShortener-Endpoints.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+});
+
+router.post('/createURL', async (req, res) => {
+  const { url, alias } = req.body;
+  let jsonbody = {};
+  if(alias != null) {
+    jsonbody = { 'url': url, 'alias': alias };
+  } else {
+    jsonbody = { 'url': url };
+  }
```
Comment (evanugarte) [inline api/main_endpoints/routes/URLShortener-Endpoints.js:24]:
```suggestion
  let jsonbody = { url, alias: alias || null };
```

File: api/main_endpoints/routes/URLShortener-Endpoints.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
+const express = require('express');
+const axios = require('axios');
+const router = express.Router();
+
+let URL_SHORTENER_BASE_URL = 'http://localhost:8000';
+
+router.get('/listAll', async (req, res) => {
```
Comment (evanugarte) [inline api/main_endpoints/routes/URLShortener-Endpoints.js:7]: all of these functions should require a token from the user, if you dont know how to do this lmk

File: src/APIFunctions/URLShortener.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+      status.responseData = err;
+      status.error = true;
+    });
+  return status;
+}
+
+export async function createURL(url, alias = null) {
+  let status = new ApiResponse();
+  await axios
+    .post(URL_SHORTENER_URL + '/URLShortener-Endpoints/createURL', { 'url': url, 'alias': alias })
```
Comment (evanugarte) [inline src/APIFunctions/URLShortener.js:23]:
```suggestion
    .post(URL_SHORTENER_URL + '/URLShortener-Endpoints/createURL', { url, alias })
```

File: src/Routing.js [hunk 5]
```diff
@@ -131,7 +132,8 @@ export default function Routing({ appProps }) {
     { Component: GoogleLoginDiscord, path: '/discordSJSU/LoginWithGoogle/:id' },
     { Component: DiscordSJSU, path: '/discordSJSU' },
     { Component: AboutPage, path: '/about'},
-    { Component: ProjectsPage, path: '/projects'}
+    { Component: ProjectsPage, path: '/projects'},
+    { Component: URLShortenerPage, path: '/s' }
```
Comment (evanugarte) [inline src/Routing.js:136]: this page should not be public, only visible if the user is an officer or admin

File: api/main_endpoints/routes/URLShortener-Endpoints.js [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    return res.sendStatus(FORBIDDEN);
+  } else if (!checkIfTokenValid(tokenReq, membershipState.OFFICER)) {
+    return res.sendStatus(UNAUTHORIZED);
+  }
+  try {
+    const response = await axios.get(URL_SHORTENER_BASE_URL + '/list');
+    const data = response.data;
+    res.json(data);
+  } catch (err) {
+    res.status(500).json({ error: 'Failed to list URLs' });
```
Comment (evanugarte) [inline api/main_endpoints/routes/URLShortener-Endpoints.js:33]: can we return the same response code as what cleezy returns? just to keep it clean

File: src/Routing.js [hunk 7]
```diff
@@ -122,6 +123,12 @@ export default function Routing({ appProps }) {
       allowedIf: userIsOfficerOrAdmin,
       redirect: '/',
       inAdminNavbar: true
+    },
+    {
+      Component: URLShortenerPage,
+      path: '/s',
```
Comment (evanugarte) [inline src/Routing.js:129]: `/s` is for talking to cleezy from the internet, lets do something else like

```suggestion
      path: '/short',
```

File: src/Pages/URLShortener/URLShortener.js [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+import Header from '../../Components/Header/Header';
+import './URLShortener-page.css';
+import { getAllURLs, createURL, deleteURL } from '../../APIFunctions/URLShortener';
+import { Container, Button, Row, Col, Input } from 'reactstrap';
+
+export default function URLShortenerPage(props) {
+  const [URL, setURL] = useState();
+  const [alias, setAlias] = useState();
+  const [allURLs, setAllURLs] = useState([]);
+  const [deleteAlias, setDeleteAlias] = useState();
```
Comment (evanugarte) [inline src/Pages/URLShortener/URLShortener.js:12]: add 2 more hooks here called `loading` and `error` which defaults to true and null respectively

in `getURLsFromDB` set loading to `false` after we get a response from the backend.

if there is an error, set the error hook to whatever the `URLsFromDB.error` is

then in the react code, only return the `<Container>` and children if `error` is null and `loading` is false

File: src/Pages/URLShortener/URLShortener-page.css [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+.url-table thead tr {
+  background-color: #3a5f7f;
+  color: #ffffff;
+  text-align: left;
+  font-weight: bold;
+}
+
+th, td {
```
Comment (evanugarte) [inline src/Pages/URLShortener/URLShortener-page.css:64]: we need to scope all this css, see #1256 on what happens if we dont

if you need help on this lmk

File: src/Pages/URLShortener/URLShortener.js [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+      return true;
+    } else if(response.error === 409) {
+      setAliasTaken(true);
+      return false;
+    } else {
+      return false;
+    }
+  }
+
+  async function checkValidURL(URL) {
```
Comment (evanugarte) [inline src/Pages/URLShortener/URLShortener.js:49]: i link this naming more
```suggestion
  async function maybeSubmitUrl(URL) {
```

File: src/Pages/URLShortener/URLShortener.js [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+    const regex =
+      /^(http(s)?:\/\/)[-a-zA-Z0-9@:%._+~#=]{2,256}\.[a-z]{2,6}\b([-a-zA-Z0-9@:%_.+~#?&//=]*)$/;
+    if (regex.test(URL)) {
```
Comment (evanugarte) [inline src/Pages/URLShortener/URLShortener.js:52]: lets avoid all caps variable names unless they are enums, constant and global or env vars

File: src/Pages/URLShortener/URLShortener.js [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+                  'Link',
+                  ''
+                ].map((element, index) => {
+                  return <th key={index}>{ element }</th>;
+                })}
+              </tr>
+            </thead>
+
+            <tbody>
+              {allURLs.map((URL, index) => {
```
Comment (evanugarte) [inline src/Pages/URLShortener/URLShortener.js:137]: anytime you see `URL` change it to lowercase, for use in snake case do
`allUrls`

**Review by evanugarte — APPROVED** [review]
https://www.youtube.com/watch?v=IZ2l_qkmGWk

**Review by Luceium — APPROVED** [review]

**Review by danknessdra — APPROVED** [review]
LGTM commented offline

**Pull request merged**